File: core/mill-metadata-core/src/main/java/io/qpointz/mill/metadata/domain/core/StructuralFacet.java
package io.qpointz.mill.metadata.domain.core;

import io.qpointz.mill.metadata.domain.AbstractFacet;
import io.qpointz.mill.metadata.domain.MetadataFacet;
import io.qpointz.mill.metadata.domain.ValidationException;
import lombok.Data;
import lombok.EqualsAndHashCode;

import java.time.Instant;

/**
 * Structural facet - physical schema binding.
 * Entity-bound facet attached to a single entity.
 */
@Data
@EqualsAndHashCode(callSuper = true)
public class StructuralFacet extends AbstractFacet {
    
    private String physicalName;
    private String physicalType;  // VARCHAR, INTEGER, etc.
    private Integer precision;
    private Integer scale;
    private Boolean nullable;
    private Boolean isPrimaryKey;
    private Boolean isForeignKey;
    private Boolean isUnique;
    private String backendType;  // jdbc, calcite, etc.
    private TableType tableType;  // For tables
    private Instant lastSyncedAt;
    
    @Override
    public String getFacetType() {
        return "structural";
    }
    
    @Override
    public void validate() throws ValidationException {
        if (physicalName == null || physicalName.isEmpty()) {
            throw new ValidationException("StructuralFacet: physicalName is required");
        }
    }
    
    @Override
    public MetadataFacet merge(MetadataFacet other) {
        if (!(other instanceof StructuralFacet)) {
            return this;
        }
        StructuralFacet otherFacet = (StructuralFacet) other;
        
        // Merge: other takes precedence for most fields
        if (otherFacet.physicalName != null) {
            this.physicalName = otherFacet.physicalName;
        }
        if (otherFacet.physicalType != null) {
            this.physicalType = otherFacet.physicalType;
        }
        if (otherFacet.precision != null) {
            this.precision = otherFacet.precision;
        }
        if (otherFacet.scale != null) {
            this.scale = otherFacet.scale;
        }
        if (otherFacet.nullable != null) {
            this.nullable = otherFacet.nullable;
        }
        if (otherFacet.isPrimaryKey != null) {
            this.isPrimaryKey = otherFacet.isPrimaryKey;
        }
        if (otherFacet.isForeignKey != null) {
            this.isForeignKey = otherFacet.isForeignKey;
        }
        if (otherFacet.isUnique != null) {
            this.isUnique = otherFacet.isUnique;
        }
        if (otherFacet.backendType != null) {
            this.backendType = otherFacet.backendType;
        }
        if (otherFacet.tableType != null) {
            this.tableType = otherFacet.tableType;
        }
        if (otherFacet.lastSyncedAt != null) {
            this.lastSyncedAt = otherFacet.lastSyncedAt;
        }
        
        return this;
    }
}

File: metadata/mill-metadata-core/src/main/kotlin/io/qpointz/mill/metadata/domain/core/StructuralFacet.kt
package io.qpointz.mill.metadata.domain.core

import io.qpointz.mill.metadata.domain.AbstractFacet
import io.qpointz.mill.metadata.domain.MetadataFacet
import io.qpointz.mill.metadata.domain.ValidationException
import java.time.Instant

/** Physical storage and schema-related attributes for table/column entities. */
open class StructuralFacet(
    var physicalName: String? = null,
    var physicalType: String? = null,
    var precision: Int? = null,
    var scale: Int? = null,
    var nullable: Boolean? = null,
    var isPrimaryKey: Boolean? = null,
    var isForeignKey: Boolean? = null,
    var isUnique: Boolean? = null,
    var backendType: String? = null,
    var tableType: TableType? = null,
    var lastSyncedAt: Instant? = null
) : AbstractFacet() {

    override val facetType: String get() = "structural"

    override fun validate() {
        if (physicalName.isNullOrEmpty()) {
            throw ValidationException("StructuralFacet: physicalName is required")
        }
    }

    override fun merge(other: MetadataFacet): MetadataFacet {
        if (other !is StructuralFacet) return this
        other.physicalName?.let { physicalName = it }
        other.physicalType?.let { physicalType = it }
        other.precision?.let { precision = it }
        other.scale?.let { scale = it }
        other.nullable?.let { nullable = it }
        other.isPrimaryKey?.let { isPrimaryKey = it }
        other.isForeignKey?.let { isForeignKey = it }
        other.isUnique?.let { isUnique = it }
        other.backendType?.let { backendType = it }
        other.tableType?.let { tableType = it }
        other.lastSyncedAt?.let { lastSyncedAt = it }
        return this
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is StructuralFacet) return false
        return physicalName == other.physicalName && physicalType == other.physicalType
    }

    override fun hashCode(): Int = (physicalName?.hashCode() ?: 0) * 31 + (physicalType?.hashCode() ?: 0)
}
